contracts/command-batcher/src/contract.rs

```diff
@@ -1,44 +1,481 @@
+
+        let active_workers: Vec<Worker> = querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: config.service_registry.to_string(),
+            msg: to_binary(&query_msg)?,
+        }))?;
+
+        let participants = active_workers
+            .into_iter()
+            .map(Worker::try_into)
+            .collect::<Result<Vec<Participant>, service_registry::ContractError>>()
+            .map_err(
+                |err: service_registry::ContractError| ContractError::InvalidParticipants {
+                    reason: err.to_string(),
+                },
+            )?
+            .try_into()
+            .map_err(|err: axelar_wasm_std::nonempty::Error| {
+                ContractError::InvalidParticipants {
+                    reason: err.to_string(),
+                }
+            })?;
+
+        Ok(Snapshot::new(
+            block
+                .time
+                .try_into()
+                .expect("violated invariant: block time cannot be invalid"),
+            block
+                .height
+                .try_into()
+                .expect("violated invariant: block height cannot be invalid"),
+            config.signing_threshold,
+            participants,
+        ))
     }
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn query(_deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+pub fn reply(deps: DepsMut, _env: Env, reply: Reply) -> Result<Response, ContractError> {
+    match parse_reply_execute_data(reply) {
+        Ok(MsgExecuteContractResponse { data: Some(data) }) => {
+            let command_batch_id = REPLY_BATCH.load(deps.storage)?;
+
+            let session_id =
+                from_binary(&data).map_err(|_| ContractError::InvalidContractReply {
+                    reason: "invalid multisig session ID".to_string(),
+                })?;
+
+            let proof_id = ProofID::new(&command_batch_id, &session_id);
+
+            PROOF_BATCH_MULTISIG.save(deps.storage, &proof_id, &(command_batch_id, session_id))?;
+
+            Ok(Response::new().add_event(Event::ProofUnderConstruction { proof_id }.into()))
+        }
+        Ok(MsgExecuteContractResponse { data: None }) => Err(ContractError::InvalidContractReply {
+            reason: "no data".to_string(),
+        }),
+        Err(_) => {
+            unreachable!("violated invariant: replied failed submessage with ReplyOn::Success")
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
     match msg {
-        QueryMsg::GetProof { proof_id } => to_binary(&query::get_proof(proof_id)?),
+        QueryMsg::GetProof { proof_id } => to_binary(&query::get_proof(deps, proof_id)?),
     }
 }
 
 pub mod query {
     use super::*;
 
-    pub fn get_proof(_proof_id: String) -> StdResult<GetProofResponse> {
-        todo!()
+    pub fn get_proof(deps: Deps, proof_id: String) -> StdResult<GetProofResponse> {
```

Comment:
I think we first need to implement deterministic session_id, no? if we link batch->session_id there could be multiple entries of the same batch pointing to different session_ids, how would the contract know which one to get from the query? I don't think we can assume it's the last one

Reply:
Hmm I see, we could include an optional session_id value in the argument, and if it's not filled, just return a list of all matching proofs. I think it's unlikely that there will be two different sessions for the same batch, and it would make it more convenient if relayers don't have to keep track of proof id.

Reply:
task to remove proof_id: https://axelarnetwork.atlassian.net/browse/AXE-1514
task to make session_id deterministic: https://axelarnetwork.atlassian.net/browse/AXE-1513